### PR TITLE
Use defaults from `meta` package for `CREATE DATABASE`

### PR DESCRIPTION
**Pull request by jsternberg**
Instead of having the parser set the defaults, the command will set the
defaults so that the constants for that are actually used. This way we
can also identify which things the user provided and which ones we are
filling with default values.

### CODE REVIEW

File: influxql/parser_test.go [hunk 1]
```diff
@@ -2812,3 +2806,11 @@ func mustParseDuration(s string) time.Duration {
 	}
 	return d
 }
+
+func Duration(v time.Duration) *time.Duration {
```
Comment (e-dard) [inline influxql/parser_test.go]: Can you keep golint happy with some comments, or maybe even better move these to `pkg` or somewhere? Or alternatively make them unexported?

File: services/meta/client_test.go [hunk 2]
```diff
@@ -108,30 +110,33 @@ func TestMetaClient_CreateDatabaseWithRetentionPolicy(t *testing.T) {
...
 		t.Fatalf("got %v, but expected %v", err, meta.ErrRetentionPolicyConflict)
 	}
 
-	// If the rp's shard group duration is different, an error should be returned.
-	rpi2 = rpi
-	rpi2.ShardGroupDuration = rpi.ShardGroupDuration + time.Minute
-	if _, err := c.CreateDatabaseWithRetentionPolicy("db0", &rpi2); err != meta.ErrRetentionPolicyConflict {
-		t.Fatalf("got %v, but expected %v", err, meta.ErrRetentionPolicyConflict)
-	}
+	/*
```
Comment (e-dard) [inline services/meta/client_test.go]: Is this supposed to be here?

File: services/meta/data.go [hunk 3]
```diff
@@ -137,40 +137,31 @@ func (data *Data) RetentionPolicy(database, name string) (*RetentionPolicyInfo,
 // Returns an error if name is blank or if a database does not exist.
 func (data *Data) CreateRetentionPolicy(database string, rpi *RetentionPolicyInfo) error {
 	// Validate retention policy.
-	if rpi.ReplicaN < 1 {
+	if rpi.Name == "" {
```
Comment (e-dard) [inline services/meta/data.go]: The method would benefit from a `if rpi == nil { return errors.New(...) }` condition.

File: services/meta/data.go [hunk 4]
```diff
@@ -844,6 +835,91 @@ func (di *DatabaseInfo) unmarshal(pb *internal.DatabaseInfo) {
...
+	var pb internal.RetentionPolicySpec
+	if err := proto.Unmarshal(data, &pb); err != nil {
+		return err
+	}
+	s.unmarshal(&pb)
+	return nil
+}
+
 // RetentionPolicyInfo represents metadata about a retention policy.
 type RetentionPolicyInfo struct {
```
Comment (e-dard) [inline services/meta/data.go]: Is there a reason why we can't rename this to `RetentionPolicy`?

File: services/meta/data.go [hunk 5]
```diff
@@ -844,6 +835,91 @@ func (di *DatabaseInfo) unmarshal(pb *internal.DatabaseInfo) {
...
+
+// NewRetentionPolicyInfo creates a new retention policy info from the specification.
+func (s *RetentionPolicySpec) NewRetentionPolicyInfo() *RetentionPolicyInfo {
+	return DefaultRetentionPolicyInfo().Apply(s)
+}
+
+// Matches checks if this retention policy specification matches
+// an existing retention policy.
+func (s *RetentionPolicySpec) Matches(rpi *RetentionPolicyInfo) bool {
+	if s.Name != "" && s.Name != rpi.Name {
```
Comment (e-dard) [inline services/meta/data.go]: `if rpi == nil { return false }`

File: services/meta/client_test.go (hunk 2 above)
Comment (jsternberg) [inline services/meta/client_test.go]: Nope. It looks like I got distracted while working on this and forgot about the comment.

File: services/meta/data.go (hunk 3 above)
Comment (jsternberg) [inline services/meta/data.go]: This should never be passed a blank retention policy. I can add it anyway though.

File: services/meta/data.go (hunk 4 above)
Comment (jsternberg) [inline services/meta/data.go]: `RetentionPolicyInfo` or `RetentionPolicySpec`? For `Info`, it's because that's what it's named everywhere and the struct encapsulates the info for a retention policy. There's also a `DatabaseInfo` struct too. I'm not sure why it was originally done that way, but that's what's in the code in a number of different places.

For the spec, I didn't want to confuse anybody by calling it a `RetentionPolicy` since it isn't. It's a specification for how to make a retention policy.

File: services/meta/data.go (hunk 3 above)
Comment (e-dard) [inline services/meta/data.go]: I assumed as much, but It's preferable that unexpected errors arise in production than panics.

File: services/meta/data.go (hunk 5 above)
Comment (jsternberg) [inline services/meta/data.go]: It might be worth having that return true. The purpose of the method is to figure out if the specification is compatible with a current retention policy. Every specification is compatible with a retention policy that doesn't exist.

I'll also update the documentation to say that.

File: services/meta/data.go (hunk 4 above)
Comment (e-dard) [inline services/meta/data.go]: Given you've added a spec for a retention policy, why do we now need to have a `RetentionPolicyInfo`, when we can have a `RetentionPolicy`. `Info` doesn't seem to add anything any longer.

In terms of it being scattered through the code-base: 

```
$ gorename -from \"github.com/influxdata/influxdb/meta\".RetentionPolicyInfo -to RetentionPolicy
```

Maybe we should get some more input, or leave it as it is.

File: influxql/parser_test.go (hunk 1 above)
Comment (jsternberg) [inline influxql/parser_test.go]: I moved them to private functions although I think I'll move them to `pkg` in the future. I just didn't want to go through the effort right now.

File: services/meta/data.go (hunk 3 above)
Comment (jsternberg) [inline services/meta/data.go]: Fixed.

File: services/meta/data.go (hunk 4 above)
Comment (jsternberg) [inline services/meta/data.go]: I think it's likely good to update it, but I don't want to overly complicate this PR since it's a very sensitive one to begin with.

File: services/meta/data.go (hunk 5 above)
Comment (jsternberg) [inline services/meta/data.go]: I updated this by having it return false since the function should never be called with a blank retention policy anyway. They also don't "match" since a match means nothing has to be done.

**Comment by e-dard** [comment]
LGTM 👍  

I think it would be worth getting at least two more review on this PR. @jwilder @corylanou @dgnorton

**Comment by corylanou** [comment]
👍 

I would add that I would like to see more test coverage.  Specifically around `meta.CreateDatabaseWithRetentionPolicy` as most of the edge cases are currently not tested.

screenshot 2016-08-10 17 27 30
